app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt

```diff
@@ -242,19 +247,106 @@ class SpotifyAppController(context: Context, val remote: SpotifyAppRemote, val w
+					loadTemporaryPlaylistState(likedSongsStateJson, likedSongsTemporaryPlaylistStateKey)
+				}
+				if (temporaryPlaylistState == null) {
+					Log.e(TAG, "Error creating liked songs playlist, falling back to app remote API")
+					queueItems = emptyList()
+					createQueueMetadata(PlayerContext(queueUri, queueTitle, null, null))
+					return@launch
+				}
+
+				saveTemporaryPlaylistState(temporaryPlaylistState, likedSongsTemporaryPlaylistStateKey)
 
 				callback?.invoke(this@SpotifyAppController)
 			} else {
-				createQueueMetadata(PlayerContext(queueUri, "Liked Songs", null, null))
+				createQueueMetadata(PlayerContext(queueUri, queueTitle, null, null))
 			}
 		}
 	}
 
+	/**
+	 * Creates a [TemporaryPlaylistState] for the supplied playlist name, creating the playlist if it
+	 * is not present and adding the queueItems in context to it. If the playlist creation fails
+	 * then null is returned.
+	 */
+	private suspend fun createTemporaryPlaylistState(playlistName: String, appSettingsTemporaryPlaylistStateKey: AppSettings.KEYS): TemporaryPlaylistState? {
+		val queueItemsHashCode = queueItems.hashCode().toString()
+		val existingPlaylistUri = webApi.getPlaylistUri(playlistName)
+		val playlistUri: String
+		val playlistId: String
+		if (existingPlaylistUri == null) {
+			Log.d(TAG, "No user playlist for $playlistName found, creating a new one.")
+			val uri = webApi.createPlaylist(playlistName, L.MUSIC_TEMPORARY_PLAYLIST_DESCRIPTION) ?: return null
+			webApi.addSongsToPlaylist(uri.id, queueItems)
+			playlistUri = uri.uri
+			playlistId = uri.id
+		} else {
+			Log.d(TAG, "User playlist for $playlistName found, using existing playlist.")
+			playlistUri = existingPlaylistUri.uri
+			playlistId = existingPlaylistUri.id
+		}
+
+		val temporaryPlaylistState = TemporaryPlaylistState(queueItemsHashCode, playlistUri, playlistId, null)
+
+		queueUri = temporaryPlaylistState.playlistUri
+
+		loadQueueCoverart(temporaryPlaylistState, appSettingsTemporaryPlaylistStateKey)
+
+		return temporaryPlaylistState
+	}
+
+	/**
+	 * Loads a [TemporaryPlaylistState] from its serialized JSON string. If the state of the
+	 * deserialized [TemporaryPlaylistState] is not valid then its contents and hash code will be
+	 * updated.
+	 */
+	private suspend fun loadTemporaryPlaylistState(temporaryPlaylistStateJson: String, appSettingsTemporaryPlaylistStateKey: AppSettings.KEYS): TemporaryPlaylistState {
+		val queueItemsHashCode = queueItems.hashCode().toString()
+		val temporaryPlaylistState = gson.fromJson(temporaryPlaylistStateJson, TemporaryPlaylistState::class.java)
+
+		if (queueItemsHashCode != temporaryPlaylistState.hashCode) {
+			Log.d(TAG, "Previous temporary playlist state is no longer valid, updating temporary playlist contents and writing new hash code.")
+			webApi.replacePlaylistSongs(temporaryPlaylistState.playlistId, queueItems)
+			temporaryPlaylistState.hashCode = queueItemsHashCode
+		}
+
+		queueUri = temporaryPlaylistState.playlistUri
```

Comment:
I believe this is unused

Reply:
This assignment is needed here as the queueUri state variable needs to be updated with the current session's queueUri that is related to the temporary playlist after it is either created or loaded since this will be different than the actual queueUri of the user library/playlist we are creating the temporary playlist of (this should stay in sync with the our currently playing playlist since we are essentially playing the temporary playlist and not the actual Liked Songs playlist for example)

Reply:
Oh oh! Right.